### PR TITLE
Bitwise Sum protocol

### PR DESCRIPTION
**Pull request by taikiy**
This is a stacked diff. The commit needs to be reviewed is eb2777f4e8f236548643897d059403bd6963496b.

## Overview
This diff implements `BitwiseSum`, which takes two bitwise shared inputs in $\mathbb{Z}\_p$ and outputs a bitwise shared number of the sum. This is an implementation of "6.2 Bitwise Sum" described in this [paper](https://iacr.org/archive/tcc2006/38760286/38760286.pdf).

## Description
This protocol is useful for adding bitwise shared numbers, but is also a part of the bit-decomposition protocol.
screenshot_2022-11-11_14-32-14

PRE∨ (PrefixOr) #195 
BIT-LT (BitwiseLessThan) #206 
SOLVED-BITS (random bit-shares) #209 
CARRIES (Carries) #219 
BIT-ADD (BitwiseSum) **this PR**

### CODE REVIEW

File: src/protocol/boolean/bitwise_sum.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+    #[allow(clippy::many_single_char_names)]
+    async fn bitwise_sum<F: Field>(a: F, b: F) -> Result<Vec<F>, BoxError>
+    where
+        Standard: Distribution<F>,
+    {
+        let world: TestWorld = make_world(QueryId);
+        let ctx = make_contexts::<F>(&world);
+        let mut rand = StepRng::new(1, 1);
```
Comment (benjaminsavage) [inline src/protocol/boolean/bitwise_sum.rs:132]: Creating the world and contexts within this function means you're doing so once for every test. That's a lot of times. Maybe move this out and pass them in?

File: src/protocol/boolean/bitwise_sum.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            bits_to_field(&bitwise_sum(c(2_147_483_647), c(1)).await?)
+        );
+        assert_eq!(
+            c(4_294_967_290),
+            bits_to_field(&bitwise_sum(c(2_147_483_645), c(2_147_483_645)).await?)
+        );
+        assert_eq!(
+            c(0),
+            bits_to_field(&bitwise_sum(c(2_147_483_645), c(2_147_483_646)).await?)
+        );
```
Comment (benjaminsavage) [inline src/protocol/boolean/bitwise_sum.rs:208]: This one seems like a place where rather than testing specific, hard-coded examples, we could easily just generate two random numbers, compute what the sum should be, and validate that it all worked correctly. That would get us test coverage of many more corner-cases.

**Review by benjaminsavage — CHANGES_REQUESTED** [review]
Just a few suggestions for improving the tests.

File: src/protocol/boolean/bitwise_sum.rs (hunk 1 above)
Comment (taikiy) [reply src/protocol/boolean/bitwise_sum.rs:132]: If I move out `ctx`, which is an array of 3 context instances, out, I will need to `narrow()` them each time I call `bitwise_sum` helper function. I think the cost is almost the same, and easier to read this way.

Comment (benjaminsavage) [reply src/protocol/boolean/bitwise_sum.rs:132]: I don’t think those compare in cost at all. Narrowing the context makes a single struct. Creating the world takes a long time because of the key exchange to set up the PRSS

**Review by benjaminsavage — APPROVED** [review]
Lots of follow up work to do, but let’s split that out into other PRs
